File: Example/XSNetwork/ErrorHandler1.m
//
//  ErrorHandler1.m
//  XSNetwork_Example
//
//  Created by shun on 2022/4/18.
//  Copyright © 2022 shun. All rights reserved.
//

#import "ErrorHandler1.h"

@implementation ErrorHandler1

- (XSErrorHanderResult *)errorHandlerWithRequestDataModel:(XSAPIBaseRequestDataModel *)requestDataModel responseURL:(NSURLResponse *)responseURL responseObject:(id)responseObject error:(NSError *)error {
    XSErrorHanderResult *xsResult = [XSErrorHanderResult new];
    
    if (error) {
        xsResult.error = error;
        return xsResult;
    } else {
        
        if ([responseObject isKindOfClass:[NSDictionary class]]) {
            int errorCode = [[responseObject objectForKey:@"code"] intValue];
            if (errorCode == 0) {
                return xsResult;
            } else {
                NSString *message = @"请求失败";
                if ([responseObject isKindOfClass:[NSDictionary class]]){
                    message = responseObject[@"message"];
                }
                NSError *newError = [[NSError alloc] initWithDomain:NSCocoaErrorDomain code:errorCode userInfo:@{NSLocalizedDescriptionKey:message,@"data":responseObject?responseObject:@{},@"URL":responseURL.URL.absoluteString}];

                xsResult.error = newError;
                return xsResult;
            }
        } else {
            xsResult.error = error;
            return xsResult;
        }
    }
}

@end

File: Example/XSNetwork/ErrorHandler1.swift
import Foundation
import XSNetwork

class ErrorHandler1: XSAPIResponseErrorHandler {
    override func errorHandler(requestDataModel: XSAPIBaseRequestDataModel,
                               responseURL: URLResponse?,
                               responseObject: Any?,
                               error: Error?) -> XSErrorHanderResult {
        let result = XSErrorHanderResult()
        if let error = error {
            result.error = error
            return result
        }
        if let dict = responseObject as? [AnyHashable: Any] {
            let errorCode = (dict["code"] as? Int) ?? 0
            if errorCode == 0 { return result }
            var message = "请求失败"
            if let msg = dict["message"] as? String { message = msg }
            result.error = NSError(domain: NSCocoaErrorDomain, code: errorCode, userInfo: [
                NSLocalizedDescriptionKey: message,
                "data": responseObject ?? [:],
                "URL": responseURL?.url?.absoluteString ?? ""
            ])
        } else {
            result.error = error
        }
        return result
    }
}
